sql-plugin/src/main/scala/com/nvidia/spark/rapids/ParquetCachedBatchSerializer.scala

```diff
@@ -1095,32 +1105,44 @@ protected class ParquetCachedBatchSerializer extends GpuCachedBatchSerializer wi
      * relationship. Each ColumnarBatch is converted to a single ParquetCachedBatch when next()
      * is called on this iterator
      */
-    class ColumnarBatchToCachedBatchIterator extends InternalRowToCachedBatchIterator {
-      override def getIterator: Iterator[InternalRow] = {
-
-        new Iterator[InternalRow] {
-          // We have to check for null context because of the unit test
-          Option(TaskContext.get).foreach(_.addTaskCompletionListener[Unit](_ => hostBatch.close()))
+    class ColumnarBatchToCachedBatchIterator extends InternalRowToCachedBatchIterator
+        with AutoCloseable {
+      val hostBatches = new ListBuffer[ColumnarBatch]()
+
+      Option(TaskContext.get).foreach(_.addTaskCompletionListener[Unit] { _ =>
+        hostBatches.foreach { hb =>
+          try {
+            hb.close()
+          } catch {
+            case e: IllegalStateException
+              if e.getMessage.startsWith("Close called too many times ") =>
+            // swallow the exception as this must be because we exhausted the iterator
+          }
+        }
+      })
 
-          val batch: ColumnarBatch = iter.asInstanceOf[Iterator[ColumnarBatch]].next
-          val hostBatch = if (batch.column(0).isInstanceOf[GpuColumnVector]) {
-            withResource(batch) { batch =>
-              new ColumnarBatch(batch.safeMap(_.copyToHost()).toArray, batch.numRows())
-            }
-          } else {
-            batch
+      override def getIterator: Iterator[InternalRow] = new Iterator[InternalRow] {
+        val batch: ColumnarBatch = iter.asInstanceOf[Iterator[ColumnarBatch]].next
+        val hostBatch = if (batch.column(0).isInstanceOf[GpuColumnVector]) {
+          withResource(batch) { batch =>
+            new ColumnarBatch(batch.safeMap(_.copyToHost()).toArray, batch.numRows())
           }
+        } else {
+          batch
+        }
+        hostBatches += hostBatch
 
-          val rowIterator = hostBatch.rowIterator().asScala
+        val rowIterator = hostBatch.rowIterator().asScala
 
-          override def next: InternalRow = rowIterator.next
+        override def next: InternalRow = rowIterator.next
 
-          override def hasNext: Boolean = rowIterator.hasNext
+        override def hasNext: Boolean = rowIterator.hasNext
+      }
 
-        }
+      override def close(): Unit = {
+        hostBatches.foreach(_.close())
```

Comment:
This needs to guard against double-close, because now we're closing it when the iteration completes _and_ when the task completes.  That means by default the task will crash under normal operation.  One easy way to resolve this:
```suggestion
      override def close(): Unit = {
        hostBatches.foreach(_.close())
        hostBatches.clear()
```